### PR TITLE
Added RecursiveLock

### PR DESCRIPTION
**Pull request by francesco-romano**
Added also `Lock` interface (implemented by `RecursiveLock` and `Mutex`).
`LockGuard` moved to used the generic `Lock` object instead of the `Mutex`

`RecursiveLock` implemented for:
- `ACE`
- `C++11`
- `POSIX`

### CODE REVIEW

**Comment by drdanz** [comment]
The patch looks good to me, even though I believe that the name of the class should be `RecursiveMutex` instead of `RecursiveLock`, to match with the name normally used.

Also, I'm not very convinced about the name of the abstract `Lock` class, I'd rather use `Lockable` and maybe `BasicLockable`, see also [BasicLockable](http://en.cppreference.com/w/cpp/concept/BasicLockable) and [Lockable](http://en.cppreference.com/w/cpp/concept/Lockable) c++14 (i think) concepts.

**Comment by francesco-romano** [comment]
@drdanz 
I get inspiration from another framework for the names, but I think what you suggested is more coherent with the current yarp setup.
I renamed to `RecursiveMutex`.

I prefer the `Lock` name, but I changed it to `Lockable`.

PS: sorry. I was sure I pushed on my fork. And unfortunately I already rebased on master, so I had to force push on this branch.

**Comment by drdanz** [comment]
:+1:

**Comment by lornat75** [comment]
I am fine about adding RecursiveLock/RecursiveMutex (I agree with keeping consistency with names).

About the abstract interface and making lock and unlock virtual: what is your thought about performance? I am not in general very concerned about performance implications of virtual functions but these objects are used extensively.

**Comment by drdanz** [comment]
I agree with @lornat75 that this will impact performance.
C++11 does not use inheritance for these classes, and `std::lock_guard` is a template class...

From https://www.cs.cmu.edu/~gilpin/c%2B%2B/performance.html

> Virtual Functions
> =================
> Virtual functions negatively affect performance in 3 main ways:
> * The constructor of an object containing virtual functions must initialize the vptr table, which is the table of pointers to its member functions.
> * Virtual functions are called using pointer indirection, which results in a few extra instructions per method invocation as compared to a non-virtual method invocation.
> * Virtual functions whose resolution is only known at run-time cannot be inlined. (For more on inlining, see the Inlining section.
>
> Templates can be used to avoid the overhead of virtual functions by using a templated class in place of inheritance. A templated class does not use the vptr table because the type of class is known at compile-time instead of having to be determined at run-time. Also, the non-virtual methods in a templated class can be inlined.
> The cost of using virtual functions is usually not a factor in calling methods that take a long time to execute since the call overhead is dominated by the method itself. In smaller methods, for example accessor methods, the cost of virtual functions is more important.


Possible solutions:

* Change `LockGuard` to a template class (but I don't know if there is a way to keep the API compatible).
* Add a `RecursiveLockGuard` class that uses a `RecursiveMutex`

**Comment by pattacini** [comment]
I do honestly think we are in the case where calls to virtual functions in the code for `lock` and `unlock` won't have any significant impact.

To any rate, please don't change the API to these methods as seen by the user code for `Mutex`, `LockGuard` and so on. It would be painful to update the code accordingly for such a - to me - minimal improvement.

**Comment by drdanz** [comment]
This should magically work with templates, but without breaking the API and without impacting performance:

```c++
template<class T>
class LockGuardTemplate
{
    // [...]
};

typedef LockGuardTemplate<yarp::os::Mutex> LockGuard;
```

(I don't like the name `LockGuardTemplate` maybe you can come up with a better name...)

You could also add a

```c++
typedef LockGuardTemplate<yarp::os::RecursiveMutex> RecursiveLockGuard;
```

And you magically have both classes defined.

**Comment by lornat75** [comment]
Not changing the API for such a class is mandatory -- from what I understood this PR was not changing the API (if I am not wrong). 

On second thought, irrespective of performance impacts using virtual functions for these classes looks like an overkill. I do not think we need dynamic binding, static binding is enough, which is what the solution suggested by @drdanz does.

So +1 for @drdanz's template.

**Comment by pattacini** [comment]
Perfectly fine for me.

**Comment by traversaro** [comment]
@drdanz @francesco-romano are you sure that in this way the specific instantiations of the template `AbstractLockGuard<Mutex>` and `AbstractLockGuard<RecursiveMutex>` are correctly exported in the Windows DLL ? 
An easy way to verify this is to compile icub-main (that use the `LockGuard` class) in Windows with YARP compiled as a shared library.

**Comment by lornat75** [comment]
@traversaro: these are always tricky issues, but if the template is instantiated in the library it should work. Anyway it is a good check to do....

**Comment by francesco-romano** [comment]
Is there somebody who can help me on the windows check?
I imagine it won't work as it is now.
But for me it is impossible to guess the right combination of `extern`, `__declspec(dllimport)` or `__declspec(dllexport)` without a windows machine...

**Comment by lornat75** [comment]
The dllimport/dllexport calls should be present for each class that exports an interface on the dll. Expect virtual interfaces, if I remember correctly.

The windows tests machines perform tests again YARP compiled as DLL they are usually a good way to catch errors.

**Comment by lornat75** [comment]
The dllimport/dllexport calls should be present for each class that exports an interface on the dll. Expect virtual interfaces, if I remember correctly.

The windows tests machines perform tests against YARP compiled as DLL they are usually a good way to catch errors.

**Comment by drdanz** [comment]
You can use explicit template instantiation to force the compiler to export the interface on the dll.

https://msdn.microsoft.com/en-us/library/by56e477.aspx
https://anteru.net/2008/11/19/318/

I'm not sure if it is needed though, we don't use it anywhere else in yarp...

**Comment by francesco-romano** [comment]
I tried an update.
Tested on OS X.

**Comment by francesco-romano** [comment]
With my last commit I have the following warnings on VS
~~~
12>C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(68): warning C4661: 'yarp::os::AbstractLockGuard<Lockable>::AbstractLockGuard(Lockable &)' : no suitable definition provided for explicit template instantiation request
12>          with
12>          [
12>              Lockable=yarp::os::Mutex
12>          ]
12>          C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(44) : see declaration of 'yarp::os::AbstractLockGuard<Lockable>::AbstractLockGuard'
12>          with
12>          [
12>              Lockable=yarp::os::Mutex
12>          ]
12>C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(68): warning C4661: 'yarp::os::AbstractLockGuard<Lockable>::~AbstractLockGuard(void)' : no suitable definition provided for explicit template instantiation request
12>          with
12>          [
12>              Lockable=yarp::os::Mutex
12>          ]
12>          C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(49) : see declaration of 'yarp::os::AbstractLockGuard<Lockable>::~AbstractLockGuard'
12>          with
12>          [
12>              Lockable=yarp::os::Mutex
12>          ]
12>C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(69): warning C4661: 'yarp::os::AbstractLockGuard<Lockable>::AbstractLockGuard(Lockable &)' : no suitable definition provided for explicit template instantiation request
12>          with
12>          [
12>              Lockable=yarp::os::RecursiveMutex
12>          ]
12>          C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(44) : see declaration of 'yarp::os::AbstractLockGuard<Lockable>::AbstractLockGuard'
12>          with
12>          [
12>              Lockable=yarp::os::RecursiveMutex
12>          ]
12>C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(69): warning C4661: 'yarp::os::AbstractLockGuard<Lockable>::~AbstractLockGuard(void)' : no suitable definition provided for explicit template instantiation request
12>          with
12>          [
12>              Lockable=yarp::os::RecursiveMutex
12>          ]
12>          C:/Dev/robotology/yarp/src/libYARP_OS/include\yarp/os/LockGuard.h(49) : see declaration of 'yarp::os::AbstractLockGuard<Lockable>::~AbstractLockGuard'
12>          with
12>          [
12>              Lockable=yarp::os::RecursiveMutex
12>          ]
~~~

I think this is due to the missing 'extern' in VS, but that gives error

@drdanz when you have a bit of time can you help me on this?

File: src/libYARP_OS/include/yarp/os/LockGuard.h [hunk 1]
```diff
@@ -27,29 +33,43 @@ namespace yarp {
...
     
     /** Assignment operator is disabled */
-    LockGuard& operator=(const LockGuard&);
+    AbstractLockGuard& operator=(const AbstractLockGuard&);
     
-    yarp::os::Mutex& mutex; /*!< underlining mutex */
+    Lockable& lock; /*!< underlining mutex */
 };
 
+#ifndef YARP_LG_EXTERN
```
Comment (drdanz) [inline src/libYARP_OS/include/yarp/os/LockGuard.h]: The right place for these define is in `yarp/conf/api.h`, so that we can re-use it if we have to do more template explicit instantiations

**Comment by drdanz** [comment]
@francesco-romano

> I think this is due to the missing 'extern' in VS, but that gives error

I think that the issue is that the implementation of the methods for the template class are in the `.cpp` file, and that means that you will not be able to use the template for any class that is not explicitly instantiated.
I suggest you move the implementation of the functions in the `.h` file

**Comment by lornat75** [comment]
Consider though that we do not have to export the template for external use with generic types but also for the ones we explicitly instantiated it for. For this the implementation can be in cpp files, right?

**Comment by drdanz** [comment]
Yes, but it will cause the warnings :)
Also if someone tries to use it, it might be confusing if the documentation is not explaining this.

**Comment by francesco-romano** [comment]
@drdanz 
My idea was to forbid the instantiation of `LockGuard` different from the two I explicitly created.
The `extern` was just to "document" the supported types

File: src/libYARP_OS/include/yarp/os/LockGuard.h (hunk 1 above)
Comment (francesco-romano) [inline src/libYARP_OS/include/yarp/os/LockGuard.h]: I looked for this definition in that file, but I did not find it. I found the contrary (extern for windows, nothing for linux)

**Comment by lornat75** [comment]
@drdanz: to prevent external use, the template should then be in the yarp namespace (as opposed to yarp/os).

**Comment by drdanz** [comment]
@lornat75 I suppose you mean you mean in the `yarp::os::impl` namespace?
Anyway, I don't think this can work... The specific instantiations cannot be exported in the `yarp::os` namespace, because they are just typedefs, they will be exported as the original class name, therefore `yarp::os::impl`

Headers in  `yarp/os` must not include headers from `yarp/os/impl`.
The definition of `LockGuard` must include the file that contains the declaration of the `AbstrackLockGuard` template class, or the methods will not be known to the user.
So, if I'm not missing something, I'd say that it is impossible to have AbstractLockGuard in `yarp::os::impl`and `LockGuard` in `yarp::os` without including headers from `yarp/os/impl` in `yarp/os`

I don't think it is worth spending time in trying to forbid the external use of the abstract class, neither templates nor inheritance will allow you to do that. The only option would be hiding the abstract class in the implementation, but then you will have to write the implementation twice (one for each class) and then the abtract class no longer makes no sense.

So, in conclusion, I'd just say make it public... If someone wants to use such a class, he probably has a good reason to do it, so I don't see why forbidding it.



Also please rename all the `XXXLockImpl` files to `XXXRecursiveLockImpl` and `RecursiveLockImpl` to `RecursiveMutexImpl` in order to avoid confusion.

**Comment by lornat75** [comment]
@drdanz you are right, I was thinking to inheritance but you are right it would not be possible without including impl header files.

Anyway it depends on how difficult it is to export the template, nothing against doing it.

**Comment by francesco-romano** [comment]
I updated the PR: I removed the .cpp file and moved the implementation in the .h

**Comment by drdanz** [comment]
Looks good to me now, except for a few very long comment lines... Anyway this PR has been here for too long, not worth stopping it for this reason. :smile:
If anyone else has anything to say about this PR, speak now, or I'm going to merge it in a few hours...

**Comment by francesco-romano** [comment]
:tada:
